### PR TITLE
PIL-1097 Privacy Policy / Terms of Service update

### PR DESCRIPTION
**Pull request by divyaBacancy**
https://user-images.githubusercontent.com/66996635/123625760-ba8f1800-d82d-11eb-8784-18d009b752a4.mov

### CODE REVIEW

File: src/services/cms/FetchPrivacyTermsDocument.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License along
+    with this program; if not, write to the Free Software Foundation, Inc.,
+    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+/* eslint-disable i18next/no-literal-string */
```
Comment (poocart) [inline src/services/cms/FetchPrivacyTermsDocument.js:20]: Not needed
```suggestion
```

File: src/components/Modals/HTMLContentModal/RenderHTMLfromPrimisic.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
+/* eslint-disable i18next/no-literal-string */
```
Comment (poocart) [inline src/components/Modals/HTMLContentModal/RenderHTMLfromPrimisic.js:1]: Can we move it to method scope?

File: src/components/Modals/HTMLContentModal/HTMLContentModal.js [hunk 3]
```diff
@@ -18,228 +18,132 @@
...
+import HTML from 'react-native-render-html';
 import get from 'lodash.get';
 
 // components
 import Spinner from 'components/Spinner';
-import ContainerWithHeader from 'components/Layout/ContainerWithHeader';
 import Modal from 'components/Modal';
+import HeaderBlock from 'components/HeaderBlock';
+import { Container } from 'components/modern/Layout';
+import type { ScrollToProps } from 'components/Modals/SlideModal';
```
Comment (poocart) [inline src/components/Modals/HTMLContentModal/HTMLContentModal.js:31]: Move to types block 🙏

File: src/components/Modals/HTMLContentModal/HTMLContentModal.js [hunk 4]
```diff
@@ -18,228 +18,132 @@
...
+  const scrollViewRef = React.useRef(null);
+  const modalRef = React.useRef();
+  const [isPrimisicDocumentFetched, setIsPrimisicDocumentFetched] = React.useState(false);
+  const [documentHTMLData, setDocumentHTMLData] = React.useState('');
+  const [scrollOffset, setScrollOffset] = React.useState(0);
+  const [containerHeight, setContainerHeight] = React.useState(0);
+  const [contentContainerHeight, setContentContainerHeight] = React.useState(0);
+
+  React.useEffect(() => {
+    async function fetchPrimisicData() {
```
Comment (poocart) [inline src/components/Modals/HTMLContentModal/HTMLContentModal.js:59]: Can we move most of logic (fetching, mapping) to `services/prismic` and then only use the output for state here?

File: src/components/Modals/HTMLContentModal/HTMLContentModal.js [hunk 5]
```diff
@@ -18,228 +18,132 @@
...
-  handleOnContentSizeChange = (width: number, height: number) => {
-    this.setState({ contentContainerHeight: height });
+  const handleOnContentSizeChange = (width: number, height: number) => {
+    setContentContainerHeight(height);
   };
+  const animationInTiming = 400;
+  const animationOutTiming = 400;
+
+  const scrollOffsetMax =
+    contentContainerHeight > 0 && containerHeight ? contentContainerHeight - containerHeight : undefined;
```
Comment (poocart) [inline src/components/Modals/HTMLContentModal/HTMLContentModal.js:106]:
```suggestion
  const scrollOffsetMax = contentContainerHeight > 0 && containerHeight
    ? contentContainerHeight - containerHeight
    : undefined;
```

**Comment by poocart** [comment]
Can we keep `HTMLContentModal` or rename it to something like `PrismicDocumentModal`? Because now `HTMLContentModal` only works with Prismic documents which takes other HTML loading out of component scope, wdyt?

File: src/services/cms/FetchPrivacyTermsDocument.js (hunk 1 above)
Comment (divyaBacancy) [reply src/services/cms/FetchPrivacyTermsDocument.js]: Done.

File: src/components/Modals/HTMLContentModal/RenderHTMLfromPrimisic.js (hunk 2 above)
Comment (divyaBacancy) [reply src/components/Modals/HTMLContentModal/RenderHTMLfromPrimisic.js]: Done.

File: src/components/Modals/HTMLContentModal/HTMLContentModal.js (hunk 3 above)
Comment (divyaBacancy) [reply src/components/Modals/HTMLContentModal/HTMLContentModal.js]: Done.

File: src/components/Modals/HTMLContentModal/HTMLContentModal.js (hunk 5 above)
Comment (divyaBacancy) [reply src/components/Modals/HTMLContentModal/HTMLContentModal.js]: Done.

File: src/components/Modals/HTMLContentModal/HTMLContentModal.js (hunk 4 above)
Comment (divyaBacancy) [reply src/components/Modals/HTMLContentModal/HTMLContentModal.js]: Done.

**Comment by divyaBacancy** [comment]
@poocart Thanks for reviewing. Done with the changes.

File: src/components/Modals/PrismicDocumentModal/PrismicDocumentModal.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  }, [prismicDocumentId]);
+
+  const handleModalClose = () => {
+    if (modalRef.current) modalRef.current.close();
+  };
+
+  const handleModalScrollTo = (p: ScrollToProps) => scrollViewRef?.current?.scrollTo(p);
+
+  const handleContentOnScroll = (event: Object) => {
+    const contentOffsetY = get(event, 'nativeEvent.contentOffset.y');
```
Comment (poocart) [inline src/components/Modals/PrismicDocumentModal/PrismicDocumentModal.js:72]:
```suggestion
    const contentOffsetY = event?.nativeEvent?.contentOffset?.y;
```

File: src/components/Modals/PrismicDocumentModal/PrismicDocumentModal.js [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    fetchPrismicData();
+  }, [prismicDocumentId]);
+
+  const handleModalClose = () => {
+    if (modalRef.current) modalRef.current.close();
+  };
+
+  const handleModalScrollTo = (p: ScrollToProps) => scrollViewRef?.current?.scrollTo(p);
+
+  const handleContentOnScroll = (event: Object) => {
```
Comment (poocart) [inline src/components/Modals/PrismicDocumentModal/PrismicDocumentModal.js:71]: You can import `ScrollEvent` from `utils/types/react-native`.

File: src/components/Modals/PrismicDocumentModal/PrismicDocumentModal.js [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    const contentOffsetY = get(event, 'nativeEvent.contentOffset.y');
+    setScrollOffset(contentOffsetY);
+  };
+
+  const handleContentOnLayout = (event: Object) => {
```
Comment (poocart) [inline src/components/Modals/PrismicDocumentModal/PrismicDocumentModal.js:76]: You can import `LayoutEvent` from `utils/types/react-native`.

File: src/components/Modals/PrismicDocumentModal/PrismicDocumentModal.js [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    >
+      <Container>
+        <HeaderBlock noBack rightItems={[{ close: true }]} onClose={handleModalClose} noPaddingTop />
+        {!isPrismicHTMLFetched && (
+          <ActivityIndicatorWrapper>
+            <Spinner />
+          </ActivityIndicatorWrapper>
+        )}
+        {!!isPrismicHTMLFetched && (
+          // $FlowFixMe: react-native types
```
Comment (poocart) [inline src/components/Modals/PrismicDocumentModal/PrismicDocumentModal.js:111]: Just curious what's the problem here why flow fails?

File: src/components/Modals/PrismicDocumentModal/RenderHTMLfromPrismic.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+      return `<h4>${text}</h4>`;
+    case 'heading6':
+      return `<h6>${text}</h6>`;
+    case 'paragraph':
+      return `<p>${text}</p>`;
+    case 'list-item':
+      return `<li>${text}</li>`;
+    default:
+      return '';
+  }
```
Comment (poocart) [inline src/components/Modals/PrismicDocumentModal/RenderHTMLfromPrismic.js:36]:
```suggestion
  /* eslint-disable i18next/no-literal-string */
  switch (type) {
    case 'heading1':
      return `<h1>${text}</h1>`;
    case 'heading4':
      return `<h4>${text}</h4>`;
    case 'heading6':
      return `<h6>${text}</h6>`;
    case 'paragraph':
      return `<p>${text}</p>`;
    case 'list-item':
      return `<li>${text}</li>`;
    default:
      return '';
  }
  /* eslint-enable i18next/no-literal-string */
```

File: src/screens/Menu/Menu.js [hunk 11]
```diff
@@ -252,11 +253,19 @@ const Menu = ({
               </FooterBanner>
             )}
             <LinksSection>
-              <LegalTextLink onPress={() => openLegalModal(ENDPOINTS.TERMS_OF_SERVICE)}>
+              <LegalTextLink
+                onPress={() =>
+                  openLegalModal(firebaseRemoteConfig.getString(REMOTE_CONFIG.PRISMIC_TERMS_OF_POLICY_DOCUMENT_ID))
```
Comment (poocart) [inline src/screens/Menu/Menu.js:258]: To make it easier to read maybe you can put `firebaseRemoteConfig.getString(REMOTE_CONFIG.PRISMIC_TERMS_OF_POLICY_DOCUMENT_ID)` as variable and have one liner like `onPress={() => openLegalModal(prismicTermsOfPolicyDocumentId)}`. Same for `PRISMIC_PRIVACY_POLICY_DOCUMENT_ID` below, wdyt?

File: src/screens/NewProfile/NewProfile.js [hunk 12]
```diff
@@ -258,50 +262,51 @@ const NewProfile = ({
     ? { default: true, floating: true, transparent: true }
     : { centerItems: [{ title: t('auth:title.chooseUsername') }] };
 
-  const openLegalModal = (endpoint: string) => Modal.open(() => <HTMLContentModal htmlEndpoint={endpoint} />);
+  const openLegalModal = (prismicDocumentId: string) =>
+    Modal.open(() => <PrismicDocumentModal prismicDocumentId={prismicDocumentId} />);
```
Comment (poocart) [inline src/screens/NewProfile/NewProfile.js:266]:
```suggestion
  const openLegalModal = (
    prismicDocumentId: string,
  ) => Modal.open(() => <PrismicDocumentModal prismicDocumentId={prismicDocumentId} />);
```

File: src/screens/NewProfile/NewProfile.js [hunk 13]
```diff
@@ -258,50 +262,51 @@ const NewProfile = ({
...
+              onPress={() => setHasAgreedToPolicy(!hasAgreedToPolicy)}
+              small
+              lightText
+              checked={hasAgreedToPolicy}
+            >
+              <CheckboxText>
+                {t('auth:withLink.readUnderstandAgreeTo', {
+                  linkedText: t('auth:privacyPolicy'),
+                  onPress: () =>
+                    openLegalModal(firebaseRemoteConfig.getString(REMOTE_CONFIG.PRISMIC_PRIVACY_POLICY_DOCUMENT_ID)),
```
Comment (poocart) [inline src/screens/NewProfile/NewProfile.js:301]: Same as per suggestion on `src/screens/Menu/Menu.js`

File: src/services/prismic.js [hunk 14]
```diff
@@ -45,3 +50,23 @@ export type Document<T> = {
 export function queryDocumentsByType<T>(type: string, options?: QueryOptions): Response<T> {
   return prismicClient.query(Prismic.Predicates.at(DOCUMENT_TYPE, type), options);
 }
+
+export async function queryDocumentsByID(id: string): Promise<Object> {
```
Comment (poocart) [inline src/services/prismic.js:54]: It looks like the result is string?
```suggestion
export async function queryDocumentsByID(id: string): Promise<string> {
```

File: src/components/Modals/PrismicDocumentModal/RenderHTMLfromPrismic.js [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License along
+    with this program; if not, write to the Free Software Foundation, Inc.,
+    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+export const renderHTMLfromPrismic = (type: string, text: string) => {
```
Comment (poocart) [inline src/components/Modals/PrismicDocumentModal/RenderHTMLfromPrismic.js:21]:
```suggestion
export const renderHTMLfromPrismic = (type: string, text: string): string => {
```

File: src/services/prismic.js [hunk 16]
```diff
@@ -45,3 +50,23 @@ export type Document<T> = {
...
+  const fetchPrismicDocumentResponse = await prismicClient.getByID(id);
+  const document = fetchPrismicDocumentResponse.data;
+  const prismicHTMLContent = [];
+  document?.title?.map((documentData) => {
+    if (!documentData.text) return null;
+    return prismicHTMLContent.push(renderHTMLfromPrismic(documentData.type, documentData.text));
```
Comment (poocart) [inline src/services/prismic.js:60]: This repeats, maybe we can extract as util?

**Comment by divyaBacancy** [comment]
@poocart Thanks for reviewing. Done with the changes.

File: src/utils/prismic.js [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+
+    You should have received a copy of the GNU General Public License along
+    with this program; if not, write to the Free Software Foundation, Inc.,
+    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+// Components
+import { renderHTMLfromPrismic } from 'components/Modals/PrismicDocumentModal/RenderHTMLfromPrismic';
+
+export const mapFromDocumentDataToString = (document: Array<Object>, prismicContent: Array<Object>) => {
```
Comment (poocart) [inline src/utils/prismic.js:24]: `prismicContent` is array of strings, Can we also create a type for document array elements? Also method return type missing

```suggestion
export const mapFromDocumentDataToString = (document: Array<Object>, prismicContent: Array<string>) => {
```

Comment (divyaBacancy) [reply src/utils/prismic.js]: @poocart Thanks. Done with the changes.

**Review by poocart — APPROVED** [review]

**Review by poocart — APPROVED** [review]

**Pull request merged**